dask/array/tests/test_array_core.py

```diff
@@ -3915,6 +3915,17 @@ def test_setitem_1d():
         dx[index] = 1
 
 
+def test_setitem_masked():
+    # Test np.ma.masked assignment to object-type arrays
+    x = np.ma.array(["a", 1, 3.14], dtype=object)
+    dx = da.from_array(x.copy(), chunks=2)
+
+    x[...] = np.ma.masked
+    dx[...] = np.ma.masked
+
+    assert_eq(x.mask, da.ma.getmaskarray(dx))
```

Comment:
I notice that if I try to do `assert_eq(x, dx)` it fails, but I would have expected not to. 

```python
>>> x
masked_array(data=[--, --, --],
             mask=[ True,  True,  True],
       fill_value='?',
            dtype=object)
>>> dx 
dask.array<setitem, shape=(3,), dtype=object, chunksize=(2,), chunktype=numpy.MaskedArray>

>>>dx.compute() 
masked_array(data=[--, --, --],
             mask=[ True,  True,  True],
       fill_value='?',
            dtype=object)           
```
not sure why the `assert_eq(x, dx)` does not work. 
But with `assert_eq(x.mask, da.ma.getmaskarray(dx))`
we are asserting the mask is the same, but not the resulting array like in all the other tests. 

@jrbourbeau Do you have any thoughts?

Reply:
Hi - I think it's because `np.ma.allclose` (https://github.com/dask/dask/blob/2022.10.2/dask/array/utils.py#L177) 
treats masked elements as being equal to co-located non-masked elements:
```python
>>> import numpy as np
>>> a = np.ma.arange(3)
>>> b = np.ma.arange(3)
>>> a[0] = np.ma.masked
>>> b[1] = np.ma.masked
>>> print(a)
[-- 1 2]
>>> print(b)
[0 -- 2]
>>> np.ma.allclose(a, b, masked_equal=True)
True
```

Reply:
... I should also say that the test is checking that both arrays (np and dask) are masked everywhere, so unless we check the mask arrays themselves, we could get a false positive. Thanks.

Reply:
Agreed, we should check that the arrays are the same. Hence me asking why trying `assert_eq(x, dx)`  fails.  We should find a way to check for this, ideally without having to compute the dask array.

Reply:
Hi again, just checking that if there's anything outstanding on this issue? It'd be nice to see it in a January release, if possible :). Thanks.